server/src/main/java/org/elasticsearch/index/mapper/DataStreamTimestampFieldMapper.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.index.IndexableField;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Map;
+
+public class DataStreamTimestampFieldMapper extends MetadataFieldMapper {
+
+    public static final String NAME = "_timestamp";
+
+    public static class Defaults  {
+
+        public static final FieldType DATA_STREAM_TIMESTAMP_FIELD_TYPE = new FieldType();
+
+        static {
+            DATA_STREAM_TIMESTAMP_FIELD_TYPE.setIndexOptions(IndexOptions.NONE);
+            DATA_STREAM_TIMESTAMP_FIELD_TYPE.freeze();
+        }
+    }
+
+    // For now the field shouldn't be useable in searches.
+    // In the future it should act as an alias to the actual data stream timestamp field.
```

Comment:
I wonder what this would look like though. It's very similar to the alias field so I wonder what @jtibshirani thinks of how this should work ?

Reply:
I think it would be similar to the code in `FieldTypeLookup` that checks `FieldAliasMapper#path()`. I think we can maybe piggyback on that.

Reply:
The field alias framework is narrow in scope and unfortunately isn't very extensible. @romseygeek is working on some refactors that will let field mappers delegate to other field types to expose search functionality. I think it will be cleaner and more forward-looking to collaborate on that.

Stepping back though, I don't know if this timestamp mapper should serve as a search-time alias for the delegate field. To me it'd be fine if it were just a hidden mapper that performed validation and recorded the data stream timestamp in the mappings. Generally we lean heavily on ECS to provide a consistent set of field names across individual schemas, and ECS has a convenient `@timestamp` field at the top level. Users can also define field aliases to allow for searching across different data streams. Does the 'timestamp' field need special treatment?

One advantage I can think of is that we could allow the delegate timestamp field to change for future indices in a data stream, and searches keep working on `_timestamp`. This doesn't seem like a common case though.

Reply:
> Generally we lean heavily on ECS to provide a consistent set of field names across individual schemas, and ECS has a convenient @timestamp field at the top level. Users can also define field aliases to allow for searching across different data streams. Does the 'timestamp' field need special treatment?

In this case, just using `@timestamp` is imo preferable and sorting by meta field `_timestamp` would not be needed.

> One advantage I can think of is that we could allow the delegate timestamp field to change for future indices in a data stream, and searches keep working on _timestamp. 

I think this becomes more useful, in cases existing indices get imported into a data stream, which is something that we would want to support down the line. 

However for now, I think there is no need for this, as the initial targeted users use ECS. I think we will be looking into this somewhere in the not so distant future.